cirq-core/cirq/ops/mixed_unitary_channel.py

```diff
@@ -0,0 +1,118 @@
+from typing import Any, Dict, Iterable, Tuple, Union
+import numpy as np
+
+from cirq import linalg, protocols, value
+from cirq._compat import proper_repr
+from cirq.ops import raw_types
+
+
+class MixedUnitaryChannel(raw_types.Gate):
+    """A generic mixture that can record the index of its selected operator.
+
+    This type of object is also referred to as a mixed-unitary channel.
+
+    Args:
+        mixture: a list of (probability, unitary) pairs
+        key: an optional measurement key string for this mixture. Simulations
+            which select a single unitary to apply will store the index
+            of that unitary in the measurement result list with this key.
+        validate: if True, validate that `mixture` describes a valid mixture.
+            This validation can be slow; prefer pre-validating if possible.
+    """
+
+    def __init__(
+        self,
+        mixture: Iterable[Tuple[float, np.ndarray]],
+        key: Union[str, value.MeasurementKey, None] = None,
+        validate: bool = False,
+    ):
+        mixture = list(mixture)
+        if not mixture:
+            raise ValueError('MixedUnitaryChannel must have at least one unitary.')
+        if not protocols.approx_eq(sum(p[0] for p in mixture), 1):
+            raise ValueError('Unitary probabilities must sum to 1.')
+        m0 = mixture[0][1]
+        num_qubits = np.log2(m0.size) / 2
+        if not num_qubits.is_integer():
+            raise ValueError(
+                f'Input mixture of shape {m0.shape} does not represent an operator over qubits.'
+            )
+        self._num_qubits = int(num_qubits)
+        for i, op in enumerate(p[1] for p in mixture):
+            if not op.size == m0.size:
+                raise ValueError(
+                    f'Inconsistent unitary sizes: op[0]: {m0.size}, op[{i}]: {op.size}'
+                )
+            if validate and not linalg.is_unitary(op):
+                raise ValueError(f'Element {i} of mixture is non-unitary.')
+        self._mixture = mixture
+        if not isinstance(key, value.MeasurementKey) and key is not None:
+            key = value.MeasurementKey(key)
+        self._key = key
+
+    @staticmethod
+    def from_mixture(
+        mixture: 'protocols.SupportsMixture', key: Union[str, value.MeasurementKey, None] = None
+    ):
+        """Creates a copy of a mixture with the given measurement key."""
+        return MixedUnitaryChannel(mixture=list(protocols.mixture(mixture)), key=key)
+
+    def __eq__(self, other) -> bool:
+        if not isinstance(other, MixedUnitaryChannel):
```

Comment:
What should be the result of equality check on a `KrausChannel` and a `MixedUnitaryChannel` that define the same underlying quantum channel?

Reply:
I'd be hesitant to return true for such a comparison. The two may have the same effect on the circuit (and thus these should return true for the `equivalent_behavior` method described above), but there are minor differences between the objects themselves: `KrausChannel.has_mixture()` is false even if the channel describes a mixture, and `MixedUnitaryChannel` has better performance due to not needing to calculate operator probabilities.

Reply:
Good point. I like the `equivalent_behavior` idea.